### PR TITLE
Impl string conversion traits for `CommandString`

### LINKED ISSUE #1135: `CommandString` has inherent `try_from` function instead of trait
Since `TryFrom` is available we should change it to convert any stringly type and provide `try_from_static(s: &'static str)` for the static use case.

Said another way, `TryFrom<T>` where `T` is `String`, `&str`, and `Box<str>`. And also `FromStr`.

### PR DESCRIPTION
**Pull request by Kixunil**
After MSRV bump `try_from` usually refers to `TryFrom` method but
`CommandString` used inherent one making it confusing and non-idiomatic.

This implements `FromStr` and `TryFrom<{stringly type}>` for
`CommandString` and deprecates the inherent method in favor of those.
To keep the code using `&'static str` efficient it provides
`try_from_static` inherent method that only converts from
`&'static str`.

Closes #1135

### CODE REVIEW

**Comment by apoelstra** [comment]
I'm tempted to just drop the existing `try_from`, impl `TryFrom<Cow<'static, str>`, and impl'ing the other `TryFrom`s off of that.

Yes, this would be a breaking change in a single major rev, but it's so slight (since we have explicitly covered all the common examples of `Into<Cow<'static, str>>`) that I think we can get away with it.

**Comment by Kixunil** [comment]
I did consider it however if we ever want to provide `TryFrom<Cow<'a, str>>` (non-static lifetime) all `'static` uses will start to silently allocate. Specialization on `'static` lifetime is impossible and will not be possible anytime soon since lifetime generics don't monomorphise. I did think about these things in the past and came to conclusion that providing `_static` methods is the least bad approach. It also expresses the intent cleanly.

I was also thinking about removing the method instead of deprecating/ Chose deprecation because it is nicer to people. I'd only keep it for one release - we could remove it from `master` immediately after release.

**Comment by Kixunil** [comment]
Added `since` to `deprecated`

**Comment by apoelstra** [comment]
Ok, I understand. I thought the `_static` method was purely a contortion to allow deprecation. But if it's to avoid specializing on `'static` which is impossible, that makes sense.

**Review by sanket1729 — APPROVED** [review]
utACK 405be52f5c83c79349f6808b395cf29316393af6

File: src/network/message.rs [hunk 1]
```diff
@@ -46,8 +47,25 @@ impl CommandString {
...
+
+    /// Convert `&'static str` to `CommandString`
+    ///
+    /// This is more efficient for string literals than non-static conversions because it avoids
+    /// allocation.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if and only if the string is
+    /// larger than 12 characters in length.
```
Comment (tcharding) [inline src/network/message.rs:63]:
```suggestion
    /// Returns an error if, and only if, the string is larger than 12 characters in length.
```

File: src/network/message.rs [hunk 2]
```diff
@@ -46,8 +47,25 @@ impl CommandString {
     ///
     /// Returns an error if and only if the string is
     /// larger than 12 characters in length.
+    #[deprecated(note = "Use `TryFrom::try_from` or `CommandString::try_from_static`", since = "0.29.0")]
     pub fn try_from<S: Into<Cow<'static, str>>>(s: S) -> Result<CommandString, CommandStringError> {
-        let cow = s.into();
+        Self::try_from_static_cow(s.into())
+    }
+
+    /// Convert `&'static str` to `CommandString`
```
Comment (tcharding) [inline src/network/message.rs:55]:
```suggestion
    /// Converts `&'static str` to a `CommandString`.
```

**Review by tcharding — COMMENTED** [review]
Super anal nits, sorry :) Totally fine to leave till another day, I'm due to do more docs cleanups at some stage soon anyways.

**Review by tcharding — APPROVED** [review]
ACK 405be52f5c83c79349f6808b395cf29316393af6

File: src/network/message.rs (hunk 1 above)
Comment (Kixunil) [reply src/network/message.rs:63]: #828

File: src/network/message.rs (hunk 2 above)
Comment (Kixunil) [reply src/network/message.rs:55]: #828

**Pull request merged**